### PR TITLE
refactor/use Subscribtion Stream

### PR DESCRIPTION
**Pull request by nalchevanidze**

### CODE REVIEW

File: src/Data/Morpheus/Types/Internal/Subscription.hs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+disconnect:: Scope 'Ws e m -> Input 'Ws -> m ()
+disconnect WS { update }  (Init clientID)  = update (delete clientID)
+
+
+-- | shared GraphQL state between __websocket__ and __http__ server,
+-- stores information about subscriptions
+data Store e m = Store 
+  { readStore :: m (ClientStore e m)
+  , writeStore :: (ClientStore e m -> ClientStore e m) -> m ()
+  }
```
Comment (theobat) [inline src/Data/Morpheus/Types/Internal/Subscription.hs:71]: So if, say, my PubSub store is RabbitMQ or Redis, this is the interface it has to respect to be able to replace the previous inmemory store ?

File: src/Data/Morpheus/Types/Internal/Subscription/ClientStore.hs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+      . snd
+      ) . HM.toList
+
+-- subscription 
+-- store
+-- stores active subsciption connections and requests
+newtype ClientStore e ( m :: * -> * ) = 
+    ClientStore 
+      { unpackStore :: HashMap ID (Client e m)
+      } deriving (Show)
```
Comment (theobat) [inline src/Data/Morpheus/Types/Internal/Subscription/ClientStore.hs:96]: This one is still a little hard to understand for me. It seems like a mapping between websocket sessionIds to something ? But I don't see why it's necessary, what is its purpose ?

File: src/Data/Morpheus/Types/Internal/Subscription.hs (hunk 1 above)
Comment (nalchevanidze) [reply src/Data/Morpheus/Types/Internal/Subscription.hs:71]: but if you want use your own monad. for example to log output. or you don't want to use default 
implementation with `MVar`  you can implement Store where you provide `write` and `read` .

File: src/Data/Morpheus/Types/Internal/Subscription/ClientStore.hs (hunk 2 above)
Comment (nalchevanidze) [reply src/Data/Morpheus/Types/Internal/Subscription/ClientStore.hs:96]: it is not sessionID. it is client ID to remove client after the Connection ended.

File: src/Data/Morpheus/Types/Internal/Subscription.hs (hunk 1 above)
Comment (theobat) [reply src/Data/Morpheus/Types/Internal/Subscription.hs:77]: So with the current implementation I cannot use an external pubsub service ? (that's what I'm after, but I don't understand everything in the current subscription design).
Say I have a subscribe function in haskell which listen to an external data store. I'd like to be able to do something like :
```haskell

subscription :: IO ()
subscription = do
  subscribeExternalService callback
  forever $ threadDelay maxBound  

callback :: SomeDataType -> IO ()
callback someDataToSendToClient = sendToClient someDataToSendToClient

```

Do you think this is possible with the current design ?

File: src/Data/Morpheus/Types/Internal/Subscription/ClientStore.hs (hunk 2 above)
Comment (theobat) [reply src/Data/Morpheus/Types/Internal/Subscription/ClientStore.hs]: But why do you need to track the clients server side ? Why is the open client-server connection not enough ?

File: src/Data/Morpheus/Types/Internal/Subscription.hs (hunk 1 above)
Comment (nalchevanidze) [reply src/Data/Morpheus/Types/Internal/Subscription.hs:77]: the store only remembers local server  connections( with subscriptions). there is no point to use any external store for it. 

if you want synchronize it with external store. you can forward received event from external source to current server and trigger it with  
`publishEventWith`.

File: src/Data/Morpheus/Types/Internal/Subscription/ClientStore.hs (hunk 2 above)
Comment (theobat) [reply src/Data/Morpheus/Types/Internal/Subscription/ClientStore.hs]: Oh ok I understand now, yes ClientConnectionStore is probably better. So the server has to track the current open connections with all its clients and that's the purpose of ClientConnectionStore.

File: src/Data/Morpheus/Types/Internal/Subscription.hs (hunk 1 above)
Comment (theobat) [reply src/Data/Morpheus/Types/Internal/Subscription.hs:77]: That makes sense, thanks !

Comment (nalchevanidze) [reply src/Data/Morpheus/Types/Internal/Subscription.hs]: @theobat 
finally i decided to define simple websocketApp

```hs
(wsApp, publish) <- webSocketsApp api
-- that can be used in 
httpAppWithEffect publish api
```
less code and less complexity for user.

what do you think?

Comment (theobat) [reply src/Data/Morpheus/Types/Internal/Subscription.hs]: Yeah I think it's great, I still have to wrap my head around all the subscription stuff it's not entirely clear to me yet. I'll try to add haddock comments to the code cause to document my digging (I'll start with this PR merged).

**Pull request merged**